### ISSUE TITLE
Register custom model / loops outside of the VISSL directory

### ISSUE THREAD

**Issue by pclucas14** [facebookresearch/vissl#211]
## 🚀 Feature

I would like to be able to integrate custom code, sur as custom model head / trunks, or custom training loops, **without having to put them directly in the vissl directory**

## Motivation & Examples

Like many users, I like to download my packages via pip or conda. Right now I have to manually copy my new models to the vissl directory within my environment. The main issue I have with this is that I have to manually keep track of my added files should I do a vissl update. 

## Note

There is already something similar already established for datasets, where I can register a new one outside of the main codebase. If there would exist something like this for models that would be great.

**Comment by prigoyal** [comment]
Hi @pclucas14 , the new components can be registered by importing registeries of vissl https://github.com/facebookresearch/vissl/blob/master/vissl/trainer/train_steps/__init__.py#L19 for example importing this function. However, the current VISSL packages don't expose the register functions in `__all__` method. But we welcome contributions :) In the next VISSL release in few weeks, we will look into incorporating this feature request :)

**Comment by pclucas14** [comment]
ok great :) thanks for the heads up

**Comment by pclucas14** [comment]
Along similar lines, is it possible to add new hooks without modifying the [hook init file](https://github.com/facebookresearch/vissl/blob/962be15e8baf43b073ccd9168bafee7727b203d0/vissl/hooks/__init__.py) ?

**Comment by prigoyal** [comment]
yep! we will open all the register functions along with the ones mentioned here! :)

**Comment by pclucas14** [comment]
Great :) Sorry to add yet another point to this, but one thing to keep in mind for future PRs would be to ensure the compatibility with the distributed slurm launcher. 

Currently I added new import lines in my version of `run_distributed_engines.py` which takes care of the registration. However when using the slurm launcher, `launch_distributed` from `vissl/utils/distributed_launcher.py` is called, which bypasses these imports. An easy hack for now is to modify `distributed_launcher.py`. 

Thanks again for the great support!
